renderdoc/driver/d3d12/d3d12_device_wrap.cpp

```diff
@@ -822,14 +1143,101 @@ HRESULT WrappedID3D12Device::CreateFence(UINT64 InitialValue, D3D12_FENCE_FLAGS
 HRESULT WrappedID3D12Device::CreateQueryHeap(const D3D12_QUERY_HEAP_DESC *pDesc, REFIID riid,
                                              void **ppvHeap)
 {
-  return m_pDevice->CreateQueryHeap(pDesc, riid, ppvHeap);
+  if(ppvHeap == NULL)
+    return m_pDevice->CreateQueryHeap(pDesc, riid, nullptr);
+
+  if(riid != __uuidof(ID3D12QueryHeap))
+    return E_NOINTERFACE;
+
+  ID3D12QueryHeap *real = NULL;
+  HRESULT ret = m_pDevice->CreateQueryHeap(pDesc, riid, (void **)&real);
+
+  if(SUCCEEDED(ret))
+  {
+    SCOPED_LOCK(m_D3DLock);
+
+    WrappedID3D12QueryHeap *wrapped = new WrappedID3D12QueryHeap(real, this);
+
+    if(m_State >= WRITING)
+    {
+      SCOPED_SERIALISE_CONTEXT(CREATE_QUERY_HEAP);
+      Serialise_CreateQueryHeap(pDesc, riid, (void **)&wrapped);
+
+      D3D12ResourceRecord *record = GetResourceManager()->AddResourceRecord(wrapped->GetResourceID());
+      record->type = Resource_QueryHeap;
+      record->Length = 0;
+      wrapped->SetResourceRecord(record);
+
+      record->AddChunk(scope.Get());
+
+      {
```

Comment:
Same again, I think query heaps definitely shouldn't be dirty-tracked. I'm not even sure it would be feasible to fetch and restore their state..